zerver/webhooks/newrelic/view.py

```diff
@@ -1,169 +1,264 @@
+
+
+def parse_payload(payload: WildValue) -> dict[str, str | None]:
+    """
+    Parses the payload and turn it into a dictionary of key-value
+    pairs that can be used to format the message body.
+    """
+    TIME_STAMP_STR: dict[str, str | None] = {
+        key: get_iso_timestamp_str(payload, value) for key, value in TIME_STAMPS_MAP.items()
+    }
 
-TOPIC_TEMPLATE = """{policy_name} ({incident_id})""".strip()
+    priority = payload.get("priority").tame(check_string_in(PRIORITIES.keys()))
+    priority_symbol = PRIORITIES.get(priority, ":alert:")
+    conditions_list = payload.get("alertConditionNames", ["Unknown condition"]).tame(
+        check_list(check_string)
+    )
+    conditions_str = ", ".join([f"**`{c}`**" for c in conditions_list])
+    policy_list = payload.get("alertPolicyNames", ["Unknown policy"]).tame(check_list(check_string))
+    alert_policy_str = ", ".join([f"**`{p}`**" for p in policy_list])
 
-# Once old is EOLed, delete old and keep new
-OLD_EVENT_TYPES = ["closed", "acknowledged", "open"]
-NEW_EVENT_TYPES = ["created", "activated", "acknowledged", "closed"]
-ALL_EVENT_TYPES = list(set(OLD_EVENT_TYPES).union(set(NEW_EVENT_TYPES)))
+    message_context: dict[str, str | None] = {
+        "title": payload.get("title", "No title.").tame(check_string),
+        "conditions_str": conditions_str,
+        "alert_policy_str": alert_policy_str,
+        "time_created_str": TIME_STAMP_STR.get("CREATED"),
+        "time_updated_str": TIME_STAMP_STR.get("UPDATED"),
+        "incident_url": payload.get("issueUrl", DEFAULT_NEWRELIC_URL).tame(check_string),
+        "total_incidents": str(payload["totalIncidents"].tame(check_int)),
+        "state": payload.get("state", "UPDATED").tame(check_string_in(INCIDENT_STATES)),
+        "status": determine_status(payload, TIME_STAMP_STR),
+        "priority": priority,
+        "priority_symbol": priority_symbol,
+    }
 
+    return message_context
 
-@webhook_view("NewRelic", all_event_types=ALL_EVENT_TYPES)
+
+def determine_status(payload: WildValue, time_stamp_map: dict[str, str | None]) -> str:
+    """
+    This function infers what type of action was taken on the incident based on the
+    timestamps provided in the payload.
+
+    Returns the action taken as a string.
+    """
+    updated_at = time_stamp_map.get("UPDATED")
+    acknowledged_at = time_stamp_map.get("ACKNOWLEDGED")
+    created_at = time_stamp_map.get("CREATED")
+    closed_at = time_stamp_map.get("CLOSED")
+
+    if updated_at:
+        if acknowledged_at and updated_at == acknowledged_at:
+            return "ACKNOWLEDGED"
+        elif created_at and updated_at == created_at:
+            return "ACTIVATED"
+        elif closed_at and updated_at == closed_at:
+            return "CLOSED"
+        else:
+            return "UPDATED"
+    incident_title = payload.get("title", "No title.").tame(check_string)
+    raise UnsupportedWebhookEventTypeError(incident_title)
```

Comment:
The incident title isn't a New Relic webhook event type that we're not able to handle. I think this should just return `"UNKNOWN STATUS"` for that final case.

Also, it seems to do all the work parsing the payload fields and then raise an error for this. If we really feel that all payloads should have an `updatedAt` field, then we should check for it in the main view function and raise the error there. But I don't think that's the case since these payloads are so customizable.

Reply:
yes, `updatedAt` already in the `ESSENTIAL_FIELDS` list we check at `payload_has_expected_fields`

Reply:
I think we can just keep `"Updated"` as the status for the final case. Because there must have been some change or update to the issue that triggered the call. Also, `"UNKNOWN STATUS"` doesn’t flow too well with the notification wording.

![image](https://github.com/user-attachments/assets/e4f6881e-a69c-49dc-929d-3ba283de002a)